### PR TITLE
Acslopez/4.11.3.0 acs usage stop duplicating vpn user entries

### PR DESCRIPTION
**Pull request by RodrigoDLopez**
## Description

Refactor on `UsageManagerImpl.createVPNUserEvent`. Currently, the present method creates and removes an VPN user event. So this PR abstract this method into an handle, and creates news methods to create and delete VPN user events.

Additionally, the new created method prevents duplicated entries for same user, with same userId, domainId and zoneId

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 1]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
         }
     }
 
-    private void createVPNUserEvent(UsageEventVO event) {
+    /**
+     * Handle with VPN user event. The Event can be:
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1752]: Precisa melhorar aqui.
`Handles VPN user create and remove events:`

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 2]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
+     * <ul>
+     *     <li>EVENT_VPN_USER_ADD</li>
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1754]: Full qualified path to the enum

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 3]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
+     *     <li>EVENT_VPN_USER_REMOVE</li>
+     * </ul>
+     * if the event isn't add or remove. An log will be throw to notify the context
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1757]: If the event received by this method is neither add nor remove, we ignore it.

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 4]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
+                deleteUsageVpnUser(event, account);
+                break;
+            default:
+                s_logger.debug(String.format("The event [type=%s, zoneId=%s, accountId=%s, userId=%s, resourceName=%s, createDate=%s] is neither of type [%s] nor [%s]",
+                        event.getType(), zoneId, accountId, userId, event.getResourceName(), event.getCreateDate(), EventTypes.EVENT_VPN_USER_ADD, EventTypes.EVENT_VPN_USER_REMOVE));
+        }
+    }
 
+    /**
+     * Find, if exists, an Usage VPN User entry and delete those entries
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1779]: Find and delete, if exists,  usage VPN user entries.

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 5]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
+     * If no Usage VPN User was found, throw an log to notify the context
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1780]: no need for this second line here.

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 6]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
+        if (CollectionUtils.isEmpty(usageVpnUsers)) {
+            s_logger.warn(String.format("No usage entry for vpn user [%s] assigned to account [%s] domain [%s] and zone [%s] was found.",
+                    userId, accountId, domainId, zoneId));
+        }
+        if (usageVpnUsers.size() > 1) {
+            s_logger.warn(String.format("More that one usage entry for vpn user [%s] assigned to account [%s] domain [%s] and zone [%s]; marking them all as deleted.", userId,
+                    accountId, domainId, zoneId));
+        }
+        for (UsageVPNUserVO vpnUser : usageVpnUsers) {
             if (s_logger.isDebugEnabled()) {
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1799]: no need to use the `isDebugEnabled`, just log them

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 7]
```diff
@@ -1747,38 +1748,95 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
-            Account acct = _accountDao.findByIdIncludingRemoved(event.getAccountId());
-            String userName = event.getResourceName();
-            UsageVPNUserVO vpnUser = new UsageVPNUserVO(zoneId, event.getAccountId(), acct.getDomainId(), userId, userName, event.getCreateDate(), null);
+            vpnUser.setDeleted(new Date());
+            _usageVPNUserDao.update(vpnUser);
+        }
+    }
+
+    /**
+     * Create an entry for Usage VPN User.
```
Comment (rafaelweingartner) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1809]: Creates an entry for the Usage VPN User. If there is already an entry in the database with the same accountId, domainId, userId and zoneId, we do not persist a new entry.

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 8]
```diff
@@ -980,7 +981,7 @@ private void createHelperRecord(UsageEventVO event) {
         } else if (isNetworkOfferingEvent(eventType)) {
             createNetworkOfferingEvent(event);
         } else if (isVPNUserEvent(eventType)) {
-            createVPNUserEvent(event);
+            handleVpnUserEvent(event);
```
Comment (leonardoventurini) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:984]: Which one is acceptable by convention, `Vpn` or `VPN` as the predicate above? :thinking:

File: usage/src/com/cloud/usage/UsageManagerImpl.java (hunk 2 above)
Comment (RodrigoDLopez) [reply usage/src/com/cloud/usage/UsageManagerImpl.java:1754]: quando tu diz full qualified path.

é o path completo dessa forma.
com.cloud.event.EventTypes#EVENT_VPN_USER_ADD

ou 
EventTypes.EVENT_VPN_USER_ADD

File: usage/src/com/cloud/usage/UsageManagerImpl.java (hunk 1 above)
Comment (RodrigoDLopez) [reply usage/src/com/cloud/usage/UsageManagerImpl.java:1752]: done!

File: usage/src/com/cloud/usage/UsageManagerImpl.java (hunk 8 above)
Comment (RodrigoDLopez) [reply usage/src/com/cloud/usage/UsageManagerImpl.java:984]: the acceptable is 'Vpn' 
I will fix the others new lines with this predicate

File: usage/src/com/cloud/usage/UsageManagerImpl.java [hunk 9]
```diff
@@ -1747,39 +1748,93 @@ private void createNetworkOfferingEvent(UsageEventVO event) {
...
-            String userName = event.getResourceName();
-            UsageVPNUserVO vpnUser = new UsageVPNUserVO(zoneId, event.getAccountId(), acct.getDomainId(), userId, userName, event.getCreateDate(), null);
+        List<UsageVPNUserVO> usageVpnUsers = findUsageVpnUsers(accountId, zoneId, userId, domainId);
+
+        if (CollectionUtils.isEmpty(usageVpnUsers)) {
+            s_logger.warn(String.format("No usage entry for vpn user [%s] assigned to account [%s] domain [%s] and zone [%s] was found.",
+                    userId, accountId, domainId, zoneId));
+        }
+        if (usageVpnUsers.size() > 1) {
+            s_logger.warn(String.format("More that one usage entry for vpn user [%s] assigned to account [%s] domain [%s] and zone [%s]; marking them all as deleted.", userId,
```
Comment (GabrielBrascher) [inline usage/src/com/cloud/usage/UsageManagerImpl.java:1794]: More **than** one usage entry

**Pull request closed without merging**